### DISCUSSION TITLE
No automatic sync on change in desktop browsers

### DISCUSSION THREAD

**Discussion by rsnel** [dani-garcia/vaultwarden#1224] [General]
### Thanks!

Thanks for this great piece of software. I came across it while I was researching self hosted password managers. I like it much more than the official server. bitwarden_rs has more features and te setup is quite easy.

### No automatic sync on change in desktop browsers

When changing something in the vault using the web interface, the browser extensions of Chromium (version 1.47.0) and Firefox (version 1.46.2) need a manual sync before they see the change.

I have websockets enabled:
WEBSOCKET_ENABLED=true
WEBSOCKET_ADDRESS=127.0.0.1
WEBSOCKET_PORT=3012

I use the apache2 recipe from the wiki:
<VirtualHost *:443>
    SSLEngine on

    ServerName HOSTNAME

    ErrorLog /var/log/apache2/bitwarden-error.log
    CustomLog /var/log/apache2/bitwarden-access.log combined

    RewriteEngine On
    RewriteCond %{HTTP:Upgrade} =websocket [NC]
    RewriteRule /notifications/hub(.*) ws://127.0.0.1:3012/$1 [P,L]
    ProxyPass / http://127.0.0.1:8888/

    ProxyPreserveHost On
    ProxyRequests Off
    RequestHeader set X-Real-IP %{REMOTE_ADDR}s

    SSLCertificateFile /etc/letsencrypt/live/HOSTNAME/fullchain.pem
    SSLCertificateKeyFile /etc/letsencrypt/live/HOSTNAME/privkey.pem
    Include /etc/letsencrypt/options-ssl-apache.conf
</VirtualHost>

In the logfile I see (while grepping on ws::) messages like:
[ws::io][INFO] Accepted a new tcp connection from 127.0.0.1:47042

(no other messages appear in this search)

### Your environment

* Bitwarden_rs version: 1.17.0

* Install method: debian package generated using bitwarden_rs-debian
* Clients used: Chrome extension 1.47.0 and Firefox extension 1.46.2
* Reverse proxy and version: apache2-2.4.38-3+deb10u4
* Version of mysql/postgresql: mariadb-1:10.3.25-0+deb10u1 (mysql)

### Steps to reproduce

Login in webinterface and add a 'Login' while Firefox extension and Chrome extension are already open. 

Configuration changed to support MySQL/MariaDB, enable SMTP en only allow account creation from a certain domain.

### Expected behaviour

Expected result is that the new login appears in Firefox and Chrome.

### Actual behaviour

They appear only after a manual sync.

### Relevant logs

[ws::io][INFO] Accepted a new tcp connection from 127.0.0.1:47042

**Comment by BlackDex** [comment]
I haven't used Apache HTTPD my self, but if you have checked the config examples on the wiki and followed those it should be fine.
How about the other way around? Does that work?

I just tested it my self (i'm using nginx), and it seems it should work if the proxy is configured correctly.

**Comment by rsnel** [comment]
Thanks for your response.

Other way around did not work.

Ok. I will try it with nginx on a test machine, and try find the problem.

**Comment by BlackDex** [comment]
@rsnel, have you discovered the issue yet, or resolved it?

I did test it with Apache HTTPD, and it works for me using the example given at the wiki.
```apache
    RewriteEngine On
    RewriteCond %{HTTP:Upgrade} =websocket [NC]
    RewriteRule /notifications/hub(.*) ws://127.0.0.1:3012/$1 [P,L]
    ProxyPass / http://127.0.0.1:8080/
```

Also, do not forget to enable the needed modules via: `a2enmod proxy_wstunnel` and `a2enmod proxy_http`.

**Comment by rsnel** [comment]
@BlackDex Thanks for your response. Sync works between browser plugins (so a change made at one browser shows up in the other browser), but a change in the web vault does not show up. Will look into it when I have time.

**Comment by rsnel** [comment]
The issue may be related to this upstream issue: https://github.com/bitwarden/web/issues/334
